Add repository-url module and move URLRepository

tbrooks8 opened the pull request:
This is related to #22116. `URLRepository` requires `SocketPermission` 
`connect`. This commit introduces a new module called "repository-url" 
where `URLRepository` will reside. With the new module, permissions can 
be removed from core.

rjernst commented:
Is there any reason FSRepository would need to be in the same module as URL repository, or could it be separate when it is eventually moved? In which case, this can be called `repository-url`?

rjernst commented on core/src/test/java/org/elasticsearch/bwcompat/RestoreBackwardsCompatIT.java:
```diff
@@ -98,7 +91,8 @@ public void testRestoreOldSnapshots() throws Exception {
             if (VersionUtils.isSnapshot(v)) continue;  // snapshots are unreleased, so there is no backcompat yet
             if (v.isRelease() == false) continue; // no guarantees for prereleases
-            if (v.before(Version.CURRENT.minimumIndexCompatibilityVersion())) continue; // we only support versions N and N-1
+            if (v.before(Version.CURRENT.minimumIndexCompatibilityVersion()))
```
Why this change? If you are going to move the continue to the next line (breaking with the convention of the other loop shortcut statements here...), please open a block.

tbrooks8 replied:
I'm not sure. I definitely did not intend that change. I'll revert that line.

rjernst commented on modules/repositories/build.gradle:
```diff
@@ -0,0 +1,33 @@
+    classname 'org.elasticsearch.plugin.repositories.RepositoriesPlugin'
+}
+
+dependencies {
```
If there are no dependencies, you don't need a dependencies section.

rjernst commented on modules/repositories/build.gradle:
```diff
@@ -0,0 +1,33 @@
+compileJava.options.compilerArgs << "-Xlint:-unchecked,-rawtypes"
+compileTestJava.options.compilerArgs << "-Xlint:-unchecked,-rawtypes"
+
+thirdPartyAudit.excludes = [
```
Not necessary.

rjernst commented on modules/repositories/src/main/java/org/elasticsearch/plugin/repositories/RepositoriesPlugin.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    public Map<String, Repository.Factory> getRepositories(Environment env, NamedXContentRegistry namedXContentRegistry) {
+        return Collections.singletonMap(URLRepository.TYPE, (metadata) -> new URLRepository(metadata, env, namedXContentRegistry));
```
nit: the parens around `metadata` are not necessary.

tbrooks8 commented:
I'll update the PR with the requested changes.

As for the name - when @s1monw and I originally talked about it, I was going to move both to a module named "repositories". But then [this](https://github.com/elastic/elasticsearch/issues/22116#issuecomment-273675490) kind of blocked that from being a straightforward change.

So we agreed that while working on the `SocketPermissions` work I would only move the `URLRepository`. I kept the naming with the assumption that at some point we would move the `FSRepository` there also.

But there is no technical reason that I know of why they would need to be in the same module. If we want to specialize this module I can make those changes.

rjernst commented:
> But there is no technical reason that I know of why they would need to be in the same module.

Since URLRepository will require network permission, but FSRepository will not, I think that would be a good split.

jasontedor commented:
> Since URLRepository will require network permission, but FSRepository will not, I think that would be a good split.

+1

tbrooks8 commented:
@s1monw are you alright with the name change?

s1monw commented:
> @s1monw are you alright with the name change?

sure

s1monw commented:
code LGTM - let do the name change and then get this in? I also wonder if we wanna add the permission already in this PR of if you wanna add it all together? either way is find

tbrooks8 commented:
I renamed it to `url-repository`.

I also worked on fixing some integration tests. I moved a snapshot restore test that used the `URLRepository` to the new module. That test required a repository path and it looks like you can only call `randomRepoPath()` on `ESIntegTestCase` if you're an internal cluster?

In order to force that I a method I had to do [this](https://github.com/elastic/elasticsearch/pull/22752/files#diff-85c8f575c2f556c0f8fdfae7ac1e5da5R47). I'm not sure if this was the correct approach? This was the first time I was dealing with the integration test harness, so let me know if there is something I should be doing instead.

jasontedor commented:
> I renamed it to `url-repository`.

I'm really sorry to be a pain here, but I think that it should be `repository-url` (we have plugins already named `repository-*`).

tbrooks8 commented:
@jasontedor I mistyped the comment. `repository-url` is actually what I named it.

jasontedor commented:
> I mistyped the comment. repository-url is actually what I named it.

Thank you. :smile:

rjernst commented:
@tbrooks8 I don't see where your `ignoreExternalCluster()` is actually called? But ESIntegTestCase subclasses to not use an external cluster (anymore, they did long ago), so I'm confused on what instigated this added boolean?

tbrooks8 commented:
@rjernst Right [here](https://github.com/elastic/elasticsearch/blob/master/test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java#L1762). 

So when I run gradle :modules:repository-url:integTest without that , this [exception](https://github.com/elastic/elasticsearch/blob/master/test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java#L1937) is being thrown.

rjernst commented:
I think you need to rename that test. That exception is there for a reason. It means the test is running with an external cluster (which integTest does everywhere _except_ in core, where the test was pulled from). In core, tests ending in IT extend from ESIntegTestCase. For all other modules, tests ending in IT mean they are running against an external cluster (a true integ test). If you rename the suffix of that test to `Tests`, it should work like a normal core integ test.

tbrooks8 commented:
Thanks. That looks like it was my issue.

tbrooks8 commented:
@jasontedor / @rjernst It looks like the build is passing at this point (now that I have worked out integration test issues). Can you update your reviews and approve it if everything looks alright?

rjernst commented on modules/repository-url/build.gradle:
```diff
@@ -0,0 +1,32 @@
+ */
+
+esplugin {
+    description 'Module for generic repositories'
```
Should be URL repository?

tbrooks8 replied:
Yep - just fixed

jasontedor commented on modules/repository-url/build.gradle:
```diff
@@ -0,0 +1,32 @@
+}
+
+compileJava.options.compilerArgs << "-Xlint:-unchecked,-rawtypes"
+compileTestJava.options.compilerArgs << "-Xlint:-unchecked,-rawtypes"
```
Can you explain why these are needed?

tbrooks8 replied:
I just copied an existing module and made appropriate modifications when I created this. I assume those are artifacts from that and I do not know any reason those are need. I'll remove them.

jasontedor replied:
Cool, let's indeed see if we can drop them.

tbrooks8 commented:
The CI build passed. So I'll merge this.
